// @frontegg/rest-api/auth/secutiry-poilicy/index.js
import { Get, Patch, Post } from '../../fetch';
import { urls } from '../../constants';
export async function getGlobalSecurityPolicy() {
  return Get(urls.identity.configurations.v1);
}
export async function getMfaPolicy() {
  return Get(`${urls.identity.configurations.v1}/mfa-policy`);
}
export async function getVendorMfaPolicy() {
  return Get(`${urls.identity.configurations.v1}/mfa-policy/vendor`);
}
export async function saveMfaPolicy(body) {
  if (body.id) {
    return Patch(`${urls.identity.configurations.v1}/mfa-policy`, body);
  } else {
    return Post(`${urls.identity.configurations.v1}/mfa-policy`, body);
  }
}
export async function getLockoutPolicy() {
  return Get(`${urls.identity.configurations.v1}/lockout-policy`);
}
export async function saveLockoutPolicy(body) {
  if (body.id) {
    return Patch(`${urls.identity.configurations.v1}/lockout-policy`, body);
  } else {
    return Post(`${urls.identity.configurations.v1}/lockout-policy`, body);
  }
}
export async function getCaptchaPolicy() {
  try {
    return await Get(`${urls.identity.configurations.v1}/captcha-policy/public`);
  } catch {
    return null;
  }
}
export async function getPasswordHistoryPolicy() {
  return Get(`${urls.identity.configurations.v1}/password-history-policy`);
}
export async function savePasswordHistoryPolicy(body) {
  if (body.id) {
    return Patch(`${urls.identity.configurations.v1}/password-history-policy`, body);
  } else {
    return Post(`${urls.identity.configurations.v1}/password-history-policy`, body);
  }
}
export async function getPasswordConfigPolicy() {
  return Get(`${urls.identity.configurations.v1}/password`);
}
