// @frontegg/redux-store/subscriptions/index.js
import { s as subscriptionsStoreName } from '../constants-4d9682b2.js';
export { s as subscriptionsStoreName } from '../constants-4d9682b2.js';
import { createSlice, createAction, combineReducers } from '@reduxjs/toolkit';
import { takeEvery, select, put, call, delay, all } from 'redux-saga/effects';
import { ISubscriptionStatus, PaymentMethodType as PaymentMethodType$1, api, ISubscriptionCancellationPolicy, ProviderType } from '@frontegg/rest-api';

function createModuleCaseReducers() {
    return {
        setLoading: {
            prepare: (payload) => ({ payload }),
            reducer: (state, action) => (Object.assign(Object.assign(Object.assign({}, state), { loading: action.payload }), (action.payload ? { error: null } : {}))),
        },
        setError: {
            prepare: (payload) => ({ payload }),
            reducer: (state, action) => (Object.assign(Object.assign({}, state), { error: action.payload, loading: false, fetching: false })),
        },
        setState: {
            prepare: (payload) => ({ payload }),
            reducer: (state, action) => (Object.assign(Object.assign({}, state), action.payload)),
        },
    };
}
function createKeyCaseReducer(key, setState) {
    return {
        prepare: (payload) => ({ payload }),
        reducer: (state, action) => (Object.assign(Object.assign({}, state), { [key]: setState(state[key], action) })),
    };
}
function createKeyCaseLoadingReducer(key) {
    return createKeyCaseReducer(key, (state, action) => (Object.assign(Object.assign({}, state), { loading: action.payload })));
}
function createKeyCaseErrorReducer(key) {
    return createKeyCaseReducer(key, (state, action) => (Object.assign(Object.assign({}, state), { error: action.payload, loading: false })));
}

const plansInitialState = {
    loading: false,
    error: null,
    fetching: true,
    plans: [],
};
const { actions: sliceActions$2, reducer: reducer$9, name: name$8 } = createSlice({
    name: `${subscriptionsStoreName}/plans`,
    initialState: plansInitialState,
    reducers: Object.assign({}, createModuleCaseReducers()),
});
const actions$9 = Object.assign({ loadPlans: createAction(`${name$8}/loadPlans`) }, sliceActions$2);

const configInitialState = {
    loading: false,
    error: null,
    fetching: true,
    config: null,
};
const reducers$6 = Object.assign({}, createModuleCaseReducers());
const { actions: configActions, reducer: reducer$8, name: name$7, } = createSlice({
    name: `${subscriptionsStoreName}/config`,
    initialState: configInitialState,
    reducers: reducers$6,
});
const actions$8 = Object.assign({ loadPaymentConfiguration: createAction(`${name$7}/loadPaymentConfiguration`) }, configActions);

const initialSubscriptionState = {
    loading: false,
    error: null,
    fetching: true,
    cancellation: {
        loading: false,
        error: null,
    },
    renewal: {
        loading: false,
        error: null,
    },
};
const reducers$5 = Object.assign(Object.assign({}, createModuleCaseReducers()), { setCancellationLoading: createKeyCaseLoadingReducer('cancellation'), setCancellationError: createKeyCaseErrorReducer('cancellation'), setRenewalLoading: createKeyCaseLoadingReducer('renewal'), setRenewalError: createKeyCaseErrorReducer('renewal') });
const { reducer: reducer$7, actions: overviewActions$1, name: name$6, } = createSlice({
    name: `${subscriptionsStoreName}/billing/subscription`,
    initialState: initialSubscriptionState,
    reducers: reducers$5,
});
const actions$7 = Object.assign({ load: createAction(`${name$6}/loadSubscription`), cancelSubscription: createAction(`${name$6}/cancelSubscription`), renewSubscription: createAction(`${name$6}/renewSubscription`) }, overviewActions$1);

const initialBillingInformationState = {
    loading: false,
    error: null,
    fetching: true,
};
const reducers$4 = Object.assign({}, createModuleCaseReducers());
const { reducer: reducer$6, actions: overviewActions, name: name$5, } = createSlice({
    name: `${subscriptionsStoreName}/billing/information`,
    initialState: initialBillingInformationState,
    reducers: reducers$4,
});
const actions$6 = Object.assign({ loadBillingInformation: createAction(`${name$5}/loadBillingInformation`), cancelSubscription: actions$7.cancelSubscription, renewSubscription: actions$7.renewSubscription }, overviewActions);

const initialPaymentMethodState = {
    loading: false,
    error: null,
    fetching: true,
};
const reducers$3 = Object.assign(Object.assign({}, createModuleCaseReducers()), { setState: {
        prepare: (payload) => ({ payload }),
        reducer: (state, action) => (Object.assign(Object.assign({}, state), action.payload)),
    } });
const { reducer: reducer$5, actions: paymentActions, name: name$4, } = createSlice({
    name: `${subscriptionsStoreName}/billing/payment`,
    initialState: initialPaymentMethodState,
    reducers: reducers$3,
});
const actions$5 = Object.assign({ loadPaymentMethod: createAction(`${name$4}/loadPaymentMethod`), submitPaymentMethod: createAction(`${name$4}/submitPaymentMethod`), submitPaymentMethodError: createAction(`${name$4}/submitPaymentMethodError`), submitPaymentMethodSuccess: createAction(`${name$4}/submitPaymentMethodSuccess`), updatePaymentMethodBillingDetails: createAction(`${name$4}/updateBillingDetails`, (payload) => ({ payload })) }, paymentActions);

const initialInvoicesState = {
    loading: false,
    error: null,
    fetching: true,
    invoices: [],
    invoiceDownload: {
        loading: false,
        error: null
    }
};
const reducers$2 = Object.assign(Object.assign({}, createModuleCaseReducers()), { setInvoiceDownloadState: {
        prepare: (payload) => ({ payload }),
        reducer: (state, action) => (Object.assign(Object.assign({}, state), { invoiceDownload: Object.assign(Object.assign({}, state.invoiceDownload), action.payload) })),
    } });
const { reducer: reducer$4, actions: sliceActions$1, name: name$3, } = createSlice({
    name: `${subscriptionsStoreName}/billing/invoices`,
    initialState: initialInvoicesState,
    reducers: reducers$2,
});
const actions$4 = Object.assign({ loadInvoices: createAction(`${name$3}/loadInvoices`), downloadInvoice: createAction(`${name$3}/downloadInvoice`, (payload) => ({ payload })) }, sliceActions$1);

const billingInitialState = {
    information: initialBillingInformationState,
    invoices: initialInvoicesState,
    paymentMethod: initialPaymentMethodState,
    subscription: initialSubscriptionState,
};
const billingActions = {
    invoices: actions$4,
    information: actions$6,
    paymentMethod: actions$5,
    subscription: actions$7,
};
const billingReducer = combineReducers({
    invoices: reducer$4,
    information: reducer$6,
    paymentMethod: reducer$5,
    subscription: reducer$7,
});

const checkoutInitialState = {
    fetching: true,
    loading: false,
    error: null,
    confirmed: false,
};
const reducers$1 = Object.assign({}, createModuleCaseReducers());
const { actions: checkoutActions, reducer: reducer$3, name: name$2 } = createSlice({
    name: `${subscriptionsStoreName}/checkout`,
    initialState: checkoutInitialState,
    reducers: reducers$1,
});
const actions$3 = Object.assign({ loadCheckout: createAction(`${name$2}/loadCheckout`), resetCheckout: createAction(`${name$2}/resetCheckout`), submitCheckout: createAction(`${name$2}/submitCheckout`), confirmCheckout: createAction(`${name$2}/confirmCheckout`, (payload) => ({ payload })), errorCheckout: createAction(`${name$2}/errorCheckout`, (payload) => ({ payload })) }, checkoutActions);

const initialSubscriptionStripeState = {
    loading: false,
    error: null,
    cardSetupIntentSecret: null,
};
const reducers = Object.assign({}, createModuleCaseReducers());
const { reducer: reducer$2, actions: reducerActions, name: name$1, } = createSlice({
    name: `${subscriptionsStoreName}/stripe`,
    initialState: initialSubscriptionStripeState,
    reducers,
});
const actions$2 = Object.assign({ loadCustomer: createAction(`${name$1}/loadCustomer`), createCardSetupIntentSecret: createAction(`${name$1}/createCardSetupIntentSecret`) }, reducerActions);

const vendorPublicConfigInitialState = {
    loading: false,
    fetching: true,
    vendorPublicConfig: null,
};
const { actions: sliceActions, reducer: reducer$1, name, } = createSlice({
    name: `${subscriptionsStoreName}/vendorPublicConfig`,
    initialState: vendorPublicConfigInitialState,
    reducers: Object.assign({}, createModuleCaseReducers()),
});
const actions$1 = Object.assign({ loadVendorPublicConfiguration: createAction(`${name}/loadVendorPublicConfiguration`) }, sliceActions);

const initialState = {
    config: configInitialState,
    plans: plansInitialState,
    checkout: checkoutInitialState,
    billing: billingInitialState,
    stripe: initialSubscriptionStripeState,
    vendorPublicConfig: vendorPublicConfigInitialState,
};
const actions = {
    config: actions$8,
    billing: billingActions,
    plans: actions$9,
    checkout: actions$3,
    stripe: actions$2,
    vendorPublicConfig: actions$1,
};
const reducer = combineReducers({
    config: reducer$8,
    billing: billingReducer,
    plans: reducer$9,
    checkout: reducer$3,
    stripe: reducer$2,
    vendorPublicConfig: reducer$1,
});

var PaymentProvider;
(function (PaymentProvider) {
    PaymentProvider["STRIPE"] = "Stripe";
})(PaymentProvider || (PaymentProvider = {}));
var SubscriptionStatus;
(function (SubscriptionStatus) {
    SubscriptionStatus["ACTIVE"] = "ACTIVE";
    SubscriptionStatus["CANCELED"] = "CANCELED";
    SubscriptionStatus["INCOMPLETE"] = "INCOMPLETE";
    SubscriptionStatus["EXPIRED"] = "EXPIRED";
    SubscriptionStatus["TRIALING"] = "TRIALING";
})(SubscriptionStatus || (SubscriptionStatus = {}));
var SubscriptionCancellationPolicy;
(function (SubscriptionCancellationPolicy) {
    SubscriptionCancellationPolicy["AT_PERIOD_END"] = "atPeriodEnd";
})(SubscriptionCancellationPolicy || (SubscriptionCancellationPolicy = {}));

function toSubscriptionCancellation({ policy }) {
    return {
        policy: toSubscriptionCancellationPolicy(),
    };
}
function toSubscriptionCancellationPolicy(policy) {
    return SubscriptionCancellationPolicy.AT_PERIOD_END;
}
function toSubscriptionStatus(status) {
    switch (status) {
        case ISubscriptionStatus.ACTIVE:
            return SubscriptionStatus.ACTIVE;
        case ISubscriptionStatus.INCOMPLETE:
            return SubscriptionStatus.INCOMPLETE;
        case ISubscriptionStatus.CANCELED:
            return SubscriptionStatus.CANCELED;
        case ISubscriptionStatus.EXPIRED:
            return SubscriptionStatus.EXPIRED;
        case ISubscriptionStatus.TRIALING:
            return SubscriptionStatus.TRIALING;
        default:
            return SubscriptionStatus.EXPIRED;
    }
}

const subscriptionResponseMock = {
    id: 'sub_1JbhYVEwsu4qiqnnfMhYAdY6',
    externalId: 'sub_1JbhYVEwsu4qiqnnfMhYAdY6',
    startDate: '2021-09-20T08:08:51.000Z',
    status: ISubscriptionStatus.ACTIVE,
    externallyManaged: false,
    cancellation: null,
    currentPeriodStart: '2021-09-20T08:08:51.000Z',
    currentPeriodEnd: '2021-10-20T08:08:51.000Z',
    plan: {
        // TODO: fix dummy data
        slug: 'test',
    },
};
const planResponseMock = [
    {
        id: 'prod_J60fUEvI7qV1eL',
        externalId: 'prod_J60fUEvI7qV1eL',
        name: 'Premium',
        description: '',
        price: {
            id: 'price_1IToe8Ewsu4qiqnndAV76J69',
            externalId: 'price_1IToe8Ewsu4qiqnndAV76J69',
            currency: 'usd',
            amount: 10000,
        },
        slug: 'premium',
    },
    {
        id: 'prod_J60duauCpXfcur',
        externalId: 'prod_J60duauCpXfcur',
        name: 'Free',
        description: 'Totally free plan. Nothing to pay.',
        price: {
            id: 'price_1ITocfEwsu4qiqnnnBHDx9fQ',
            externalId: 'price_1ITocfEwsu4qiqnnnBHDx9fQ',
            currency: 'usd',
            amount: 0,
        },
        slug: 'free',
    },
];
const invoicesMock = [
    {
        id: 'in_1JbhYWEwsu4qiqnnrUh6hsHa',
        externalId: 'in_1JbhYWEwsu4qiqnnrUh6hsHa',
        subscriptionId: 'sub_1JbhYVEwsu4qiqnnfMhYAdY6',
        paymentDate: '2021-09-20T08:08:51.000Z',
        totalAmount: 10000,
        currency: 'usd',
        paid: true,
        receiptNumber: '',
    },
    {
        id: 'in_1JbhYWEwsu4qiqnnrUh6hsHa',
        externalId: 'in_1JbhYWEwsu4qiqnnrUh6hsHa',
        subscriptionId: 'sub_1JbhYVEwsu4qiqnnfMhYAdY6',
        paymentDate: '2021-08-20T08:08:51.000Z',
        totalAmount: 10000,
        currency: 'usd',
        paid: true,
        receiptNumber: '',
    },
];
const paymentMethodsMock = [
    {
        id: 'id',
        externalId: 'externalId',
        type: PaymentMethodType$1.CARD,
        isDefault: true,
        last4: '4242',
        expMonth: 11,
        expYear: 25,
        brand: 'visa',
        billingDetails: {
            name: 'Dummy',
            email: 'dummy@email.com',
        },
    },
];
const vendorPublicConfigurationResponseMock = {
    allowPlanDowngrade: true,
    allowPlanCancellation: true,
};

function* subscriptionSagas() {
    yield takeEvery(actions$7.load, loadSubscriptionTenant);
    yield takeEvery(actions$7.cancelSubscription, cancelSubscription);
    yield takeEvery(actions$7.renewSubscription, renewSubscription);
}
function* loadSubscriptionTenant() {
    const tenantId = yield select((state) => { var _a, _b; return (_b = (_a = state.auth) === null || _a === void 0 ? void 0 : _a.user) === null || _b === void 0 ? void 0 : _b.tenantId; });
    yield loadSummaries(tenantId);
}
function* loadSubscription() {
    yield put(actions$7.setLoading(true));
    try {
        const [subscription] = yield call(api.subscriptions.getManagedSubscriptions);
        yield put(actions$7.setState({
            subscription,
            fetching: false,
            loading: false,
            error: null,
        }));
    }
    catch (e) {
        yield put(actions$7.setError(e));
    }
}
function* cancelSubscription() {
    const { subscription } = yield select((state) => state.subscriptions.billing.subscription);
    if (!subscription) {
        yield put(actions$7.setCancellationError('Subscription not found.'));
        return;
    }
    if (subscription === null || subscription === void 0 ? void 0 : subscription.externallyManaged) {
        yield put(actions$7.setCancellationError('Billing is externally managed.'));
        return;
    }
    const { id: subscriptionId, cancellation, status } = subscription || {};
    const isCancellable = !cancellation && status === ISubscriptionStatus.ACTIVE;
    if (isCancellable) {
        try {
            yield put(actions$7.setCancellationLoading(true));
            yield call(api.subscriptions.cancelManagedSubscription, subscriptionId);
            yield loadSubscription();
            yield put(actions$7.setCancellationLoading(false));
        }
        catch (e) {
            yield put(actions$7.setCancellationError(e.message));
        }
    }
}
function* renewSubscription() {
    const { subscription } = yield select((state) => state.subscriptions.billing.subscription);
    if (!subscription) {
        yield put(actions$7.setCancellationError('Subscription not found.'));
        return;
    }
    if (subscription === null || subscription === void 0 ? void 0 : subscription.externallyManaged) {
        yield put(actions$7.setCancellationError('Billing is externally managed'));
        return;
    }
    const { id: subscriptionId, cancellation } = subscription || {};
    const renewable = (cancellation === null || cancellation === void 0 ? void 0 : cancellation.policy) === ISubscriptionCancellationPolicy.AT_PERIOD_END;
    if (renewable) {
        try {
            yield put(actions$7.setRenewalLoading(true));
            yield call(api.subscriptions.renewManagedSubscription, subscriptionId);
            yield loadSubscription();
            yield put(actions$7.setRenewalLoading(false));
        }
        catch (e) {
            yield put(actions$7.setCancellationError(e.message));
        }
    }
}
/*********************************
 *  Preview Sagas
 *********************************/
function* loadMock() {
    yield put(actions$7.setLoading(true));
    yield delay(500);
    yield put(actions$7.setState({
        subscription: subscriptionResponseMock,
        fetching: false,
        loading: false,
        error: null,
    }));
}
function* subscriptionSagasMock() {
    yield takeEvery(actions$7.load, loadMock);
}

function* subscriptionBillingInformationSagas() {
    yield takeEvery(actions$6.loadBillingInformation, loadBillingInformation);
}
function* loadBillingInformation() {
    yield loadBillingInformationAction(false);
}
function* loadBillingInformationAction(forceActive) {
    const paymentProvider = yield select((state) => { var _a; return (_a = state.subscriptions.config.config) === null || _a === void 0 ? void 0 : _a.paymentProvider; });
    const tenantId = yield select((state) => { var _a, _b; return (_b = (_a = state.auth) === null || _a === void 0 ? void 0 : _a.user) === null || _b === void 0 ? void 0 : _b.tenantId; });
    if (!paymentProvider || !tenantId) {
        yield put(actions$6.setError(!paymentProvider ? 'Internal feature failure' : 'Not authorized'));
        return;
    }
    yield loadSummaries(tenantId, forceActive);
}
function* loadSummaries(tenantId, forceActive) {
    var _a, _b;
    yield put(actions$6.setLoading(true));
    try {
        const summary = yield call(api.subscriptions.getSubscriptionSummaries, tenantId);
        const { currentPlanId, externallyManaged } = summary;
        let subscriptionResponse = null;
        let planResponse;
        if (!externallyManaged) {
            [, planResponse] = yield all([
                yield loadSubscription(),
                call(api.subscriptions.getSubscriptionPlan, currentPlanId),
            ]);
            subscriptionResponse = yield select((state) => state.subscriptions.billing.subscription.subscription || null);
        }
        else {
            planResponse = yield call(api.subscriptions.getSubscriptionPlan, currentPlanId);
        }
        yield put(actions$6.setState(Object.assign(Object.assign({ loading: false, fetching: false, summary }, (subscriptionResponse
            ? {
                subscription: {
                    id: subscriptionResponse.id,
                    externalId: subscriptionResponse.externalId,
                    startDate: subscriptionResponse.startDate,
                    currentPeriodStart: subscriptionResponse.currentPeriodStart,
                    currentPeriodEnd: subscriptionResponse.currentPeriodEnd,
                    status: forceActive ? SubscriptionStatus.ACTIVE : toSubscriptionStatus(subscriptionResponse.status),
                    cancellation: subscriptionResponse.cancellation && toSubscriptionCancellation(subscriptionResponse.cancellation),
                    trialEnd: subscriptionResponse.trialEnd ? subscriptionResponse.trialEnd : null,
                },
            }
            : {})), (planResponse
            ? {
                plan: {
                    id: planResponse.id,
                    name: planResponse.name,
                    description: planResponse.description,
                    price: ((_a = planResponse.price) === null || _a === void 0 ? void 0 : _a.amount) || 0,
                    currency: ((_b = planResponse.price) === null || _b === void 0 ? void 0 : _b.currency) || 'usd',
                    recurringInterval: 'month',
                    slug: planResponse.slug,
                },
            }
            : {}))));
    }
    catch (e) {
        yield put(actions$6.setError(e.message));
    }
}
/*********************************
 *  Preview Sagas
 *********************************/
function* loadBillingInformationMock() {
    yield loadSummariesMock();
}
function* loadSummariesMock() {
    var _a, _b;
    yield put(actions$6.setLoading(true));
    yield delay(500);
    yield put(actions$6.setState({
        loading: false,
        fetching: false,
        summary: {
            subscriptionId: subscriptionResponseMock.id,
            paymentMethodId: 'mockPaymentMethodId',
            tenantConfigurationId: 'mockTenantConfigurationId',
            providerType: ProviderType.Stripe,
            externallyManaged: false,
            currentPlanId: planResponseMock[0].id,
            defaultPlanId: planResponseMock[0].id,
        },
        plan: Object.assign(Object.assign({}, planResponseMock[0]), { price: (_a = planResponseMock[0].price) === null || _a === void 0 ? void 0 : _a.amount, currency: (_b = planResponseMock[0].price) === null || _b === void 0 ? void 0 : _b.currency, recurringInterval: 'month' }),
    }));
}
function* subscriptionBillingInformationSagasMock() {
    yield takeEvery(actions$6.loadBillingInformation, loadBillingInformationMock);
}

function* subscriptionsPaymentMethodSagas() {
    yield takeEvery(actions$5.loadPaymentMethod, loadPaymentMethod);
    yield takeEvery(actions$5.submitPaymentMethod, submitPaymentMethod);
    yield takeEvery(actions$5.submitPaymentMethodError, submitPaymentMethodError);
    yield takeEvery(actions$5.submitPaymentMethodSuccess, submitPaymentMethodSuccess);
    yield takeEvery(actions$5.updatePaymentMethodBillingDetails, updateBillingDetails);
}
function* loadPaymentMethod() {
    yield put(actions$5.setLoading(true));
    try {
        const paymentMethods = yield call(api.subscriptions.getPaymentMethods);
        const paymentMethod = paymentMethods[0];
        yield put(actions$5.setState({
            paymentMethod,
            loading: false,
            fetching: false,
        }));
    }
    catch (e) {
        yield put(actions$5.setError(e.message));
    }
}
function* updateBillingDetails({ payload, }) {
    yield put(actions$5.setLoading(true));
    const { id, email, address, callback } = payload;
    try {
        yield call(api.subscriptions.updatePaymentMethodBillingDetails, id, Object.assign({ email }, address));
        yield call(loadPaymentMethod);
        callback === null || callback === void 0 ? void 0 : callback(true);
    }
    catch (e) {
        yield put(actions$5.setError(e.message));
        callback === null || callback === void 0 ? void 0 : callback(false);
    }
    yield put(actions$5.setLoading(false));
}
function* submitPaymentMethod() {
    yield put(actions$5.setLoading(true));
}
function* submitPaymentMethodError({ payload: error }) {
    yield put(actions$5.setError(error));
}
function* submitPaymentMethodSuccess() {
    yield put(actions$5.loadPaymentMethod());
}
/*********************************
 *  Preview Sagas
 *********************************/
function* loadPaymentMethodMock() {
    yield put(actions$5.setLoading(true));
    yield delay(500);
    const paymentMethod = paymentMethodsMock[0];
    yield put(actions$5.setState({
        paymentMethod,
        loading: false,
        fetching: false,
    }));
}
function* subscriptionsPaymentMethodSagasMock() {
    yield takeEvery(actions$5.loadPaymentMethod, loadPaymentMethodMock);
}

function* subscriptionInvoicesSagas() {
    yield takeEvery(actions$4.loadInvoices, loadInvoices);
    yield takeEvery(actions$4.downloadInvoice, downloadInvoice);
}
function* loadInvoices() {
    yield put(actions$4.setLoading(true));
    try {
        const responseInvoices = yield call(api.subscriptions.getSubscriptionInvoices);
        const invoices = responseInvoices.map((invoice) => ({
            id: invoice.id,
            externalId: invoice.externalId,
            subscriptionId: invoice.subscriptionId,
            paymentDate: new Date(Date.parse(invoice.paymentDate)),
            totalAmount: +((invoice.totalAmount || 0) / 100).toFixed(2),
            currency: invoice.currency || 'usd',
            paid: invoice.paid || false,
            receiptNumber: invoice.receiptNumber,
        }));
        yield put(actions$4.setState({
            loading: false,
            fetching: false,
            invoices,
        }));
    }
    catch (e) {
        yield put(actions$4.setError(e.message));
    }
}
function* downloadInvoice({ payload }) {
    yield put(actions$4.setInvoiceDownloadState({ loading: true, error: null }));
    try {
        yield call(api.subscriptions.getSubscriptionInvoicePdf, payload.invoiceId, payload.filename);
        yield put(actions$4.setInvoiceDownloadState({ loading: false, error: null }));
    }
    catch (e) {
        yield put(actions$4.setInvoiceDownloadState({ loading: false, error: e.message || null }));
    }
}
/*********************************
 *  Preview Sagas
 *********************************/
function* loadInvoicesMock() {
    yield put(actions$4.setLoading(true));
    yield delay(500);
    const selectPlanTitle = 'Premium';
    yield put(actions$4.setState({
        loading: false,
        fetching: false,
        invoices: invoicesMock.map((invoice) => (Object.assign(Object.assign({}, invoice), { selectedPlan: selectPlanTitle, paymentDate: new Date(Date.parse(invoice.paymentDate)), totalAmount: +((invoice.totalAmount || 0) / 100).toFixed(2) }))),
    }));
    yield put(actions$4.setLoading(false));
}
function* subscriptionInvoicesSagasMock() {
    yield takeEvery(actions$4.loadInvoices, loadInvoicesMock);
}

function* billingSagas() {
    yield all([
        call(subscriptionBillingInformationSagas),
        call(subscriptionsPaymentMethodSagas),
        call(subscriptionInvoicesSagas),
        call(subscriptionSagas),
    ]);
}
function* billingSagasMock() {
    yield all([
        call(subscriptionBillingInformationSagasMock),
        call(subscriptionsPaymentMethodSagasMock),
        call(subscriptionInvoicesSagasMock),
        call(subscriptionSagasMock),
    ]);
}

function* checkoutSagas() {
    yield takeEvery(actions$3.loadCheckout, loadCheckout);
    yield takeEvery(actions$3.resetCheckout, resetCheckout);
    yield takeEvery(actions$3.submitCheckout, submitCheckout);
    yield takeEvery(actions$3.confirmCheckout, confirmPlan);
    yield takeEvery(actions$3.errorCheckout, errorCheckout);
}
function* loadCheckout() {
    yield put(actions$3.setState({
        fetching: false,
        loading: false,
        error: null,
        confirmed: false,
    }));
}
function* resetCheckout() {
    yield put(actions$3.setState({
        loading: false,
        error: null,
        confirmed: false,
    }));
}
/**
 * Based on payment provider type
 */
function* submitCheckout() {
    yield put(actions$3.setState({
        loading: true,
        error: null,
    }));
}
function* confirmPlan({ payload: { paymentMethodId, planId } }) {
    const subscription = yield select((state) => state.subscriptions.billing.subscription.subscription);
    const summary = yield select((state) => state.subscriptions.billing.information.summary);
    const isTrialing = (subscription === null || subscription === void 0 ? void 0 : subscription.status) === SubscriptionStatus.TRIALING;
    const hasPaymentMethod = !!(summary === null || summary === void 0 ? void 0 : summary.paymentMethodId);
    if (!subscription) {
        yield put(actions$3.setState({
            loading: false,
            error: 'Subscription not found',
        }));
        return;
    }
    yield put(actions$3.setState({
        loading: true,
        error: null,
    }));
    if (isTrialing && hasPaymentMethod) {
        yield confirmCheckout();
        yield put(actions$7.setState({
            subscription: Object.assign(Object.assign({}, subscription), { status: ISubscriptionStatus.ACTIVE }),
        }));
        return;
    }
    try {
        yield call(api.subscriptions.updateManagedSubscription, subscription.id, {
            paymentMethodId,
            planId,
        });
        yield all([
            loadBillingInformation(),
            loadPaymentMethod(),
            loadInvoices(),
        ]);
        yield confirmCheckout();
    }
    catch (e) {
        yield put(actions$3.setState({
            loading: false,
            error: e.message,
        }));
    }
}
function* confirmCheckout() {
    yield put(actions$3.setState({
        loading: false,
        error: null,
        confirmed: true,
    }));
}
function* errorCheckout({ payload }) {
    yield put(actions$3.setState({
        loading: false,
        error: payload,
    }));
}
/*********************************
 *  Preview Sagas
 *********************************/
function* checkoutSagasMock() {
    yield takeEvery(actions$3.resetCheckout, resetCheckout);
}

function* plansSagas() {
    yield takeEvery(actions$9.loadPlans, loadPlans);
}
function* loadPlans() {
    yield put(actions$9.setLoading(true));
    try {
        const products = yield call(api.subscriptions.getSubscriptionPlans);
        const plans = products.map((item) => {
            var _a, _b;
            return ({
                id: item.id,
                name: item.name,
                description: item.description,
                price: ((_a = item.price) === null || _a === void 0 ? void 0 : _a.amount) || 0,
                currency: ((_b = item.price) === null || _b === void 0 ? void 0 : _b.currency) || 'usd',
                recurringInterval: 'month',
                slug: item.slug,
            });
        });
        yield put(actions$9.setState({
            fetching: false,
            loading: false,
            plans,
        }));
    }
    catch (e) {
        yield put(actions$9.setError(e.message));
    }
}
/*********************************
 *  Preview Sagas
 *********************************/
function* loadPlansMock() {
    yield put(actions$9.setLoading(true));
    yield delay(500);
    yield put(actions$9.setState({
        fetching: false,
        loading: false,
        plans: planResponseMock.map((item) => {
            var _a, _b;
            return (Object.assign(Object.assign({}, item), { price: ((_a = item.price) === null || _a === void 0 ? void 0 : _a.amount) || 0, currency: ((_b = item.price) === null || _b === void 0 ? void 0 : _b.currency) || 'usd', recurringInterval: 'month' }));
        }),
    }));
}
function* plansSagasMock() {
    yield takeEvery(actions$9.loadPlans, loadPlansMock);
}

var PaymentMethodType;
(function (PaymentMethodType) {
    PaymentMethodType["UNKNWON"] = "unknown";
    PaymentMethodType["CARD"] = "card";
})(PaymentMethodType || (PaymentMethodType = {}));

function* configSagas() {
    yield takeEvery(actions$8.loadPaymentConfiguration, loadPaymentConfiguration);
}
function* loadPaymentConfiguration() {
    yield put(actions$8.setLoading(true));
    try {
        const response = yield call(api.subscriptions.getPaymentProviders) || [];
        const stripePaymentProvider = response.find((paymentProvider) => paymentProvider.status === '1' && paymentProvider.providerType === ProviderType.Stripe);
        if (stripePaymentProvider) {
            yield loadStripePaymentConfiguration();
        }
        else {
            yield put(actions$8.setError('Payment provider not configured'));
        }
    }
    catch (e) {
        yield put(actions$8.setError(e.message));
    }
}
function* loadStripePaymentConfiguration() {
    yield put(actions$8.setLoading(true));
    try {
        const response = yield call(api.subscriptions.getStripePaymentProviderConfiguration);
        yield put(actions$8.setState({
            loading: false,
            fetching: false,
            config: {
                paymentProvider: PaymentProvider.STRIPE,
                apiKey: response.publishableKey
            }
        }));
    }
    catch (e) {
        yield put(actions$8.setError(e.message));
    }
}
/*********************************
 *  Preview Sagas
 *********************************/
function* loadPaymentConfigurationMock() {
    yield put(actions$8.setLoading(true));
    yield delay(500);
    yield put(actions$8.setState({
        loading: false,
        fetching: false,
        config: {
            paymentProvider: PaymentProvider.STRIPE,
            apiKey: ''
        },
    }));
}
function* configSagasMock() {
    yield takeEvery(actions$8.loadPaymentConfiguration, loadPaymentConfigurationMock);
}

function* subscriptionStripeSagas() {
    yield takeEvery(actions$2.createCardSetupIntentSecret, createCardSetupIntentSecret);
}
function* createCardSetupIntentSecret({ payload }) {
    yield put(actions$2.setState({
        loading: true,
        error: null,
        cardSetupIntentSecret: null,
    }));
    try {
        const request = {};
        if (payload) {
            request.paymentMethodId = payload;
        }
        const { setupIntentSecret } = yield call(api.subscriptions.createStripePaymentMethodSetupIntentSecret, request);
        yield put(actions$2.setState({
            cardSetupIntentSecret: setupIntentSecret,
            loading: false,
        }));
    }
    catch (e) {
        yield put(actions$2.setError(e.message));
    }
}

function* vendorPublicConfigSagas() {
    yield takeEvery(actions$1.loadVendorPublicConfiguration, loadVendorPublicConfiguration);
}
function* loadVendorPublicConfiguration() {
    yield put(actions$1.setLoading(true));
    try {
        const vendorPublicConfig = yield call(api.subscriptions.getVendorPublicConfigurations);
        yield put(actions$1.setState({
            fetching: false,
            loading: false,
            vendorPublicConfig,
        }));
    }
    catch (e) {
        yield put(actions$1.setError(e.message));
    }
}
/*********************************
 *  Preview Sagas
 *********************************/
function* loadVendorPublicConfigurationMock() {
    yield put(actions$1.setLoading(true));
    yield delay(500);
    yield put(actions$1.setState({
        fetching: false,
        loading: false,
        vendorPublicConfig: vendorPublicConfigurationResponseMock,
    }));
}
function* vendorPublicConfigSagasMock() {
    yield takeEvery(actions$1.loadVendorPublicConfiguration, loadVendorPublicConfigurationMock);
}

function* sagas() {
    yield all([
        call(billingSagas),
        call(checkoutSagas),
        call(plansSagas),
        call(configSagas),
        call(subscriptionStripeSagas),
        call(vendorPublicConfigSagas),
    ]);
}
function* mockSagas() {
    yield all([
        call(billingSagasMock),
        call(checkoutSagasMock),
        call(plansSagasMock),
        call(configSagasMock),
        call(vendorPublicConfigSagasMock),
    ]);
}

// export store
// export store
var subscriptionsStore = {
    sagas,
    mockSagas,
    reducer,
    actions,
    initialState,
    storeName: subscriptionsStoreName,
};

export { PaymentMethodType, PaymentProvider, SubscriptionCancellationPolicy, SubscriptionStatus, subscriptionsStore as default, actions as subscriptionActions, initialState as subscriptionInitialState, reducer as subscriptionReducers, sagas as subscriptionSagas, mockSagas as subscriptionSagasMock };
