// @frontegg/admin-portal/node/LoginBoxRenderer.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var FronteggApp = require('./FronteggApp-5e722d8d.js');
var reduxStore = require('@frontegg/redux-store');
var types = require('@frontegg/types');
var index = require('./index-a24c5f7c.js');
var formik_esm = require('./formik.esm-13bb4769.js');
require('./utils.js');
require('./AppHolder.js');
require('@frontegg/rest-api');
require('uuid');

const defaultTheme = 'modern';
const getThemeByName = (name) => { var _a; return (_a = formik_esm.themes$1[name]) !== null && _a !== void 0 ? _a : formik_esm.themes$1[defaultTheme]; };
class LoginBoxRenderer {
    constructor(name, themeOptions, store, appOptions) {
        this.setRoute = () => { };
        this.themeOptions = {};
        const contextOptions = { baseUrl: 'preview' };
        this.store = store !== null && store !== void 0 ? store : reduxStore.createFronteggStore({ context: contextOptions }, this, true, undefined, undefined, true);
        this.options = Object.assign({ themeOptions, iframeRendering: true, metadata: {}, previewMode: true, builderMode: true, store: this.store, contextOptions, onRedirectTo: () => {
                console.debug('path');
            } }, appOptions);
        this.app = new FronteggApp.FronteggApp(this.options, name, true);
        types.Metadata.set({}, name);
    }
    render(loginBoxContainer) {
        let loginBoxEl = loginBoxContainer.querySelector('#root');
        if (!loginBoxEl) {
            loginBoxEl = loginBoxContainer.ownerDocument.createElement('div');
            loginBoxEl.setAttribute('id', 'root');
            loginBoxContainer.appendChild(loginBoxEl);
        }
        this.app.loginBoxContainer = loginBoxContainer;
        this.app.loginBoxEl = loginBoxEl;
        const setThemeSetter = (themeSetter) => {
            this.themeSetter = themeSetter;
        };
        const setStaticRouteSetter = (setStaticRoute) => {
            this.setRoute = setStaticRoute;
        };
        const loginBox = FronteggApp.React.createElement(index.LoginBoxPage, {
            options: this.options,
            injector: this.app,
            setThemeSetter,
            setStaticRouteSetter,
            staticRoute: '/account/login',
        });
        const wrapper = FronteggApp.React.createElement(FronteggApp.React.Suspense, { fallback: '' }, loginBox);
        FronteggApp.ReactDom.render(wrapper, loginBoxEl);
        this.app.loading = false;
    }
    unmount() {
        try {
            if (this.app.loginBoxEl) {
                FronteggApp.ReactDom.unmountComponentAtNode(this.app.loginBoxEl);
            }
        }
        catch (e) {
            console.error('Failed to unmount login box renderer', e);
        }
    }
    setTheme(themeOptions) {
        var _a;
        this.themeOptions = themeOptions;
        (_a = this.themeSetter) === null || _a === void 0 ? void 0 : _a.call(this, this.themeOptions);
    }
    setStaticRoute(staticRoute) {
        var _a;
        (_a = this.setRoute) === null || _a === void 0 ? void 0 : _a.call(this, staticRoute);
    }
    setStore(state) {
        this.store.dispatch({
            type: 'auth/setState',
            payload: state,
        });
    }
}

exports.LoginBoxRenderer = LoginBoxRenderer;
exports['default'] = LoginBoxRenderer;
exports.defaultTheme = defaultTheme;
exports.getThemeByName = getThemeByName;
